pkg/make_rpm.py

```diff
@@ -199,36 +230,103 @@ def SetupWorkdir(self, spec_file, original_dir):
         os.makedirs(dst_dir, 0o777)
       shutil.copy(os.path.join(original_dir, f), dst_dir)
 
+
     # Copy the spec file, updating with the correct version.
+
+    # Replace scriptlets if they're provided.
+    tpl_replacements = {
+      'PRE_SCRIPTLET': "%pre\n" + self.pre_scriptlet,
+      'POST_SCRIPTLET': "%post\n" + self.post_scriptlet,
+      'PREUN_SCRIPTLET': "%preun\n" + self.preun_scriptlet,
+      'POSTUN_SCRIPTLET': "%postun\n" + self.postun_scriptlet,
+    }
+
     spec_origin = os.path.join(original_dir, spec_file)
     self.spec_file = os.path.basename(spec_file)
     replacements = {}
     if self.version:
       replacements['Version:'] = self.version
     if self.release:
       replacements['Release:'] = self.release
-    CopyAndRewrite(spec_origin, self.spec_file, replacements)
+    CopyAndRewrite(spec_origin, self.spec_file,
+                   replacements=replacements,
+                   template_replacements=tpl_replacements)
+
+    # preamble replacement, %description, %install, %files -f substitution
+    self.preamble_file = None
+    if preamble_file:
+      # Copy in the various other files needed to build the RPM
+      self.preamble_file = os.path.basename(preamble_file)
+      tpl_replacements = {}
+      if self.version:
+        tpl_replacements['VERSION_FROM_FILE'] = self.version
+      if self.release:
+        tpl_replacements['RELEASE_FROM_FILE'] = self.release
+      CopyAndRewrite(os.path.join(original_dir, preamble_file),
+                     self.preamble_file,
+                     template_replacements=tpl_replacements)
+
+    self.description_file = None
```

Comment:
> This feels slightly weird, for a few reasons.
> 
> We have an RpmBuilder class, but this group of a few class members (*_file) are not initialized in the **init** method. Just to keep that clean, these options could be passed to the class initializer and pre-set there.
> `self.description_file = os.path.basename(description_file)` if description_file or None
> 
> So when we reach this point it might look like
> 
> ```python
> if self.description_file:
>     shutil.copy(self.description_file, os.getcwd())
> ```
> 
> But this is obviously different than what you have, but I can't quite understand that.

At the time, I believe was trying to be consistent with the surrounding code (see `self.spec_file`), which has some values that are not defined in the constructor.

Now that I think on this a little, all attributes in the class should probably be initialized to some default value even if they are to actually be set later, as is required by many more "static" programming languages.  This could avoid a bunch of spaghetti that would otherwise show up when you define everything.

> Maybe this is just me forgetting rpmbuild behavior, but maybe the code needs some annotation to explain why certain things are done.
>
>     * It seems we move some files to the build dir even though we include them, but we have the full paths. Can we not %include with a full path.  And maybe we are doing that.

I honestly don't know whether or not `%include` can accept absolute paths.  TBH, I never tried.

>       I see  shutil.copy(os.path.join(original_dir, install_script_file), os.getcwd())
>       which feels odd.  It seems that in the rule impl we had the full path to the install_script_file, then we basename that somewhere, then add in the path of the dir we started in. It's hard to follow.  Maybe it just needs some comments in the code about what original_dir means and the significance of were we chdir 

Comments are easy enough to do.  I'll write these in the immediate term.

However, you are very much correct in that this is a confusing mess.  There are currently two ways to inject data into the RPM spec template:

1) A shell-script-style ${VARIABLE} template, which is expanded into the RPM spec file directly
2) A rpm-style macro statement, e.g. `%include %build_rpm_install`, which are expanded inline by`rpmbuild` when it is called.

I get the feeling that this whole process would be greatly simplified if this process was broken down into two or three distinct phases:

1) The template spec file is expanded to a more fully-qualified spec file that can be invoked by `rpmbuild` with minimal additional magic.
2) The rpmbuild directory structure is created, and, if necessary (see below), the already-built artifacts are put in the right place.
3) `rpmbuild` is invoked with the necessary arguments to build the RPM (This would be `make_rpm.py`)

This could help us dramatically reduce the complexity of `make_rpm.py`.

We might be able to somewhat reduce the number of file copies this process entails by having the `%install` phase simply copy (or hardlink) the files from the bazel build tree to where they are supposed to be relative to RPM builds the package.  Haven't yet tried this out either.

Reply:
Tracked in #209.

Reply:
I elected to move the scriptlets out of the constructor and put them into `SetupWorkdir` instead.  All the values are initialized to `None` in the constructor, though.

See 9f906b39161e5092ea3ebf8143c64344b2618715.